main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,278 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+        Expression.MkClosureRef(optimize(ref).asInstanceOf[Expression.Ref], freeVars, tpe, loc)
+      case Expression.ApplyRef(sym, args, tpe, loc) => Expression.ApplyRef(sym, optimizeExps(args), tpe, loc)
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) =>
+        Expression.ApplyTail(sym, formals, optimizeExps(actuals), tpe, loc)
+      case Expression.ApplyHook(hook, args, tpe, loc) => Expression.ApplyHook(hook, optimizeExps(args), tpe, loc)
+      case Expression.Apply(exp, args, tpe, loc) => Expression.Apply(optimize(exp), optimizeExps(args), tpe, loc)
+      case Expression.Unary(op, exp, tpe, loc) => Expression.Unary(op, optimize(exp), tpe, loc)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) => optimizeBinaryExpression(e0.asInstanceOf[Expression.Binary])
+      case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+        val e1 = optimize(exp1)
+        e1 match {
+          // Elimination of dead branches (e.g. if (true) e1 else e2).
+          case Expression.True => optimize(exp2)
+          case Expression.False => optimize(exp3)
+          case _ =>
+            // Elimination of redundant branching (e.g. if(c1, if(c2, e2, e3), e3) -> if (c1 && c2, e2, e3))
+            val e2 = optimize(exp2)
+            val e3 = optimize(exp3)
+            (e2, e3) match {
+              case (Expression.IfThenElse(se1, se2, se3, _, _), _) =>
+                if (se3 == e3) Expression.IfThenElse(and(e1, se1, loc), se2, se3, tpe, loc)
+                else if (se2 == e3) Expression.IfThenElse(and(e1, not(se1, loc), loc), se3, se2, tpe, loc)
+                else Expression.IfThenElse(e1, e2, e3, tpe, loc)
+              case (_, Expression.IfThenElse(se1, se2, se3, _, _)) =>
+                if (se3 == e2) Expression.IfThenElse(and(not(e1, loc), se1, loc), se2, se3, tpe, loc)
+                else if (se2 == e2) Expression.IfThenElse(and(not(e1, loc), not(se1, loc), loc), se3, se2, tpe, loc)
+                else Expression.IfThenElse(e1, e2, e3, tpe, loc)
+              case _ => Expression.IfThenElse(e1, e2, e3, tpe, loc)
+            }
+        }
+      case Expression.Let(sym, exp1, exp2, tpe, loc) => Expression.Let(sym, optimize(exp1), optimize(exp2), tpe, loc)
+      case Expression.LetRec(sym, exp1, exp2, tpe, loc) => Expression.LetRec(sym, optimize(exp1), optimize(exp2), tpe, loc)
+      case Expression.Is(sym, tag, exp, loc) =>
+        val e = optimize(exp)
+        // Elimination of run-time tag checks for Unit.
+        if (e == Expression.Unit) Expression.False
+        // Elimination of run-time tag checks of single-valued enums.
+        else if (isSingleCaseEnum(sym)) Expression.True
+        else Expression.Is(sym, tag, e, loc)
+      case Expression.Tag(sym, tag, exp, tpe, loc) =>
+        // Remove the tag on a single-valued enum.
+        val e = optimize(exp)
+        if (isSingleCaseEnum(sym)) e else Expression.Tag(sym, tag, e, tpe, loc)
```

Comment:
The current errors are coming from here, where the `tag` and `untag` expressions are removed for single case enums.  I think I also have to write a function that changes every occurrence of the `Type` class corresponding to the enum to be updated to the type of the object the enum is wrapping.  Do you agree that this is the next step to solve this issue?